### PR TITLE
[WIP] Adds genies

### PR DESCRIPTION
**Pull request by unid15**
- Added (currently only admin-spawnable) genies - simple mobs that must be controlled by a player
- Genies take 50% brute damage, have 100 health and possess a mediocre melee attack. They can fly and they're not affected by atmos. In addition, they have knowledge of three spells: Smoke, Ethereal Jaunt and Conjure Item.
- Added Conjure Item spell. Only available to genies, it has the power to instantly create any item in the game. Some items (like high frequency katanas, admin laser guns, admin banhammers or the Goonecode) are too powerful to be created by this spell - attempting to do so will result in creation of a rabid mimic. It has a cooldown of 10 seconds.

Sprites: Credit to **Bad_Spriter**
djinni

**TODO**:
- Balance
- Add a "misc roles" preference / jobban option for genies (in the future stuff like talkie toasters could use them too)

### CODE REVIEW

**Comment by Kurfursten** [comment]
So I'm guessing they despawn after three uses of Conjure?

**Comment by Ephx1** [comment]
At the very least Conjure shouldn't work after three uses.

**Comment by Shadowmech88** [comment]
Animate the genie's tail so it looks more ethereal instead of looking like he's standing on one blue leg.
I think it'd be a good idea to log when genies conjure things.

What happens if you type an incomplete name? Is it like the sandbox spawn menu where it'll open a list of everything containing the string?

**Comment by PJB3005** [comment]
@Shadowmech88 looking from the code, you need the exact name or it won't find it.

**Comment by unid15** [comment]
Oil lamps work kinda like posibrains. Rubbing them will send a message to ghosts, ghosts may then sign up to be the genie and eventually a random ghost is picked. 

Conjure item only works three times, it does nothing afterwards. You also need to type in the exact name of the item, or it won't work. I could make it work like spawn though

**Comment by Wizardcrying** [comment]
You should let an admin allow a genie (be it through varediting or something else) to be able to uncap a genie's conjure uses.

**Comment by unid15** [comment]
Already can be done through varediting, by changing the "times_used" var to -999999

**Comment by ihadtoregisterforthis** [comment]
I can't help but to fear what will come of shitters or metagamers if they get themselves into a genie, or what will come of the server if you use a whitelist for genie role.
In any case there needs to be a genie jobban.

**Comment by mrpain666** [comment]
From an admin perspective, at the very least, create logs of everything these things conjure, please.

**Comment by Wizardcrying** [comment]
See the second picture unid posted.

**Comment by NigglyWiggly** [comment]
Can't wait until people use this to conjure HF blades and M1911-P's and shit.
Could there be exclusions to what can be spawned?

**Comment by Ephx1** [comment]
@NigglyWiggly 
"Some items (like high frequency katanas, admin laser guns, admin banhammers or the Goonecode) are too powerful to be created by this spell - attempting to do so will result in creation of a rabid mimic. It has a cooldown of 10 seconds."
Uhhhhhhhhhh

**Comment by Kurfursten** [comment]
The M1911-P seems to be okay to summon. These are the forbidden items:

/obj/item/weapon/katana/hfrequency
/obj/item/weapon/reagent_containers/glass/bottle/adminordrazine
/obj/item/weapon/reagent_containers/pill/adminordrazine
/obj/item/device/fuse_bomb/admin
/obj/item/weapon/banhammer/admin
/obj/item/weapon/melee/energy/axe
/obj/item/weapon/gun/energy/laser/admin
/obj/item/toy/gooncode
/obj/item/weapon/gun/energy/pulse_rifle

File: code/game/objects/items/devices/flashlight.dm [hunk 1]
```diff
@@ -130,6 +130,100 @@
...
+	var/list/ghost_volunteers = list()
+
+/obj/item/device/flashlight/lamp/oil/magic/rub()
+	..()
+
+	if(our_genie)
+		return
+
+	if(!collecting_ghosts)
+		usr << "<span class='sinister'>\The [src] glows in your hands and starts shaking.</span>"
```
Comment (PJB3005) [inline code/game/objects/items/devices/flashlight.dm:162]: Gonna need to change this to `to_chat(usr, ...)`.

File: code/game/objects/items/devices/flashlight.dm [hunk 2]
```diff
@@ -130,6 +130,100 @@
...
+	..()
+	spawn(10)
+	start_collecting_ghosts(usr)
+
+/obj/item/device/flashlight/lamp/oil/magic/proc/start_collecting_ghosts(mob/user)
+	search_for_candidates()
+
+	spawn(200)
+		if(!ghost_volunteers.len)
+			user << "<span class='info'>Nothing happens.</span>"
```
Comment (PJB3005) [inline code/game/objects/items/devices/flashlight.dm:175]: `to_chat(user, ...)`

File: code/game/objects/items/devices/flashlight.dm [hunk 3]
```diff
@@ -130,6 +130,100 @@
...
+			return
+		else
+			var/mob/dead/observer/O = pick(ghost_volunteers)
+			if(istype(O) && O.client && O.key)
+				spawn_genie(O, user)
+			else
+				user << "<span class='info'>Nothing happens.</span>"
```
Comment (PJB3005) [inline code/game/objects/items/devices/flashlight.dm:182]: `to_chat(usr, ...)`

File: code/game/objects/items/devices/flashlight.dm [hunk 4]
```diff
@@ -130,6 +130,100 @@
...
+/obj/item/device/flashlight/lamp/oil/magic/proc/search_for_candidates()
+	ghost_volunteers = list()
+	collecting_ghosts = 1
+
+	for(var/mob/dead/observer/O in dead_mob_list)
+		if(O.client)
+			if(O.has_enabled_antagHUD == 1 && config.antag_hud_restricted)
+				continue
+
+			O << "<span class=\"recruit\">Somebody has found a magic lamp! (<a href='?src=\ref[src];signup=\ref[O]'>Become a genie</a>)</span>"
```
Comment (PJB3005) [inline code/game/objects/items/devices/flashlight.dm:194]: `to_chat(O, ...)`

File: code/game/objects/items/devices/flashlight.dm [hunk 5]
```diff
@@ -130,6 +130,100 @@
...
+
+/obj/item/device/flashlight/lamp/oil/magic/Topic(href, href_list)
+	if(href_list["signup"])
+		var/mob/dead/observer/O = locate(href_list["signup"])
+		if(!O) return
+
+		ghost_volunteers |= O
+
+		O << "<span class='info'>You have been added to the list of possible genies.</span>"
```
Comment (PJB3005) [inline code/game/objects/items/devices/flashlight.dm:203]: `to_chat(O, ...)`

File: code/game/objects/items/devices/flashlight.dm [hunk 6]
```diff
@@ -130,6 +130,100 @@
...
+
+/obj/item/device/flashlight/lamp/oil/magic/proc/spawn_genie(mob/dead/observer/new_soul, mob/user)
+	our_genie = new(get_turf(src))
+	our_genie.visible_message("<span class='notice'>\A [our_genie] emerges out of \the [src]!</span>")
+
+	our_genie.ckey = new_soul.ckey
+	our_genie << "<b>You are a genie.</b>"
```
Comment (PJB3005) [inline code/game/objects/items/devices/flashlight.dm:210]: `to_chat(our_genie, ...)`

**Comment by ghost** [comment]
Merge when, or should I poke the sprite a bit, maybe make it slightly translucent and a bit of a smoke effect like @Shadowmech88 suggested?

**Comment by Imagebard** [comment]
Do you waste a slot in "conjure" if you write a item name wrong?

**Comment by ConkerDev** [comment]
I dont understand the limit in conjuration. Genies dont magically loose their powers after granting three wishes, they instead become free of their master. If a genie needs something to protect themselves, or do something of their own will, they should be able to. What about a system like nethack's wishing (without gods fucking you up because you are a heathen, of course)? Or, even better, wishing is done via admin fuckery, so you can ask for more abstract things like inmortality, all the gold in the world, or peace, or a piece of that spessman's ass or whaterver.

**Comment by Imagebard** [comment]
>wishing is done via admin fuckery

This is already possible in the game as-is, and depending on what you mean this request will either:
Makes genies useless when there are no admins on.
Give players admin-level powers
Either is a bad result.

**Comment by ihadtoregisterforthis** [comment]
It really depends on the source, but in Disney's Aladdin the genie could only go free if it was wished. Aladdin 2 was a mistake but Jafar doesn't say anything about going free after the wishing and revenge.
But yes, wishes should do more than spawn things. As a first iteration is great and can give feedback on wether they'll need more powers or they can manage with what they have.

**Comment by Imagebard** [comment]
What no
Are you just starting a new PR due to conflicts?

**Comment by NigglyWiggly** [comment]
That's the way I always do it!

**Comment by Imagebard** [comment]
5 days ago and it still isn't back up
He also didn't do any github contributions in the last 28 days

Shit

**Comment by Shadowmech88** [comment]
Well the code's all here. If you really wanted to you could copy it from the diff and make your own PR.

**Comment by Imagebard** [comment]
I'm not a coder you big bully

**Comment by clusterfack** [comment]
Yes you are just write some code

**Pull request closed without merging**